src/rust_option.rs

```diff
@@ -30,517 +27,240 @@ impl<T: Sized> OptionTarget for &mut T {}
 impl<T: Sized> private::Sealed for Pin<&mut T> {}
 impl<T: Sized> OptionTarget for Pin<&mut T> {}
 
-impl<T: Sized> private::Sealed for *const T {}
-impl<T: Sized> OptionTarget for *const T {}
-
-impl<T: Sized> private::Sealed for *mut T {}
-impl<T: Sized> OptionTarget for *mut T {}
-
 #[cfg(feature = "alloc")]
 impl<T: Sized> private::Sealed for Box<T> {}
 #[cfg(feature = "alloc")]
 impl<T: Sized> OptionTarget for Box<T> {}
 
-#[repr(C)]
-union OptionInner<T: OptionTarget> {
-    value: ManuallyDrop<T>,
-    empty: usize,
-}
-
-impl<T: OptionTarget> OptionInner<T> {
-    fn none() -> Self {
-        Self { empty: 0 }
-    }
-
-    #[cfg(feature = "alloc")]
-    fn new(t: T) -> Self {
-        Self {
-            value: ManuallyDrop::new(t),
-        }
-    }
-
-    fn has_value(&self) -> bool {
-        let _: () = assert_option_safe::<&T>();
-        unsafe { self.empty != 0 }
-    }
-
-    fn into_inner_unchecked(mut self) -> T {
-        let value = unsafe { ManuallyDrop::take(&mut self.value) };
-        unsafe { core::ptr::write(&mut self as _, OptionInner::none()) };
-        value
-    }
-}
-
-impl<T: OptionTarget> Drop for OptionInner<T> {
-    fn drop(&mut self) {
-        if self.has_value() {
-            unsafe { ManuallyDrop::drop(&mut self.value) };
-        }
-    }
-}
-
 // ABI compatible with C++ rust::Option<T> (not necessarily core::option::Option<T>).
 #[repr(C)]
 pub struct RustOption<T: OptionTarget> {
-    inner: OptionInner<T>,
+    repr: core::mem::MaybeUninit<T>,
```

Comment:
this does feel slightly weird, especially when combined with #[repr(C)]. This might be the core reason why I need to use indirect ABI now but I'm not sure

Reply:
I guess another question is, can I have `inner: core::mem::MaybeUninit<Option<T>>`? Then the `from` impl could be `Self { inner: MaybeUninit::new(option) }`